Show volume information for attached storage

cloudbehl opened the pull request:
Signed-off-by: Ankush Behl <cloudbehl@gmail.com>

Issue:https://jira.coreos.com/browse/CONSOLE-905

coreos-ui commented:
Can one of the admins verify this patch?

coreos-ui commented:
Can one of the admins verify this patch?

spadgett commented:
/ok-to-test

spadgett commented:
jenkins rebuild

spadgett commented:
@cloudbehl we could break this into smaller PRs where we add the volumes to the workload pages first, then add the full container page in a follow on

cloudbehl commented:
@spadgett There are three workloads (stateful sets, jobs, cron jobs) pages I am facing the Issue.
1.Stateful Sets: The Object for the Stateful set is not like all other resource (pod, deployment, deployment config). It has `volumeMounts` key but doesn't have `volumes` key inside container object. Where in all another object I was able to see it. So should I write a code specific to Stateful set?

2. (Jobs, Cron Jobs) : I have added a code to display the mounted volumes thinking the Object will be generic and will have keys like `volumes` and `volumeMounts` like all other resources I have added attachments but I was not able to test because I was not sure how to add mounted volume details to both the pages.

Related to linking from the Utilized By row to the container details page. I am not able to get to the container details page from the particular resource. So the link is broken right now with the PR. Can you help me get the container name so I can fix this?

spadgett commented:
Let's do the basic workloads first (deployment, dc, rc, rs) and leave the others for follow on PRs.

> Related to linking from the Utilized By row to the container details page. I am not able to get to the container details page from the particular resource. So the link is broken right now with the PR. Can you help me get the container name so I can fix this?

Only pods have a container page right now, so it won't just work :) This code needs to be written. For now, I'd make it so it's not a link and just show the volumes on the workload page.

This is a good candidate to break up into several PRs.

cloudbehl commented:
> Let's do the basic workloads first (deployment, dc, rc, rs) and leave the others for follow on PRs.

I have removed the code from all other workload added to the pod, deployment, dc, rc and rs. 

> Only pods have a container page right now, so it won't just work :) This code needs to be written. For now, I'd make it so it's not a link and just show the volumes on the workload page.

I thought we need to link to pod's container page from every workload. From what I understand from this is that every workload will have a container detail page. Is it correct?

spadgett commented:
> I thought we need to link to pod's container page from every workload. From what I understand from this is that every workload will have a container detail page. Is it correct?

Yes, but there is no page yet to link to. The existing pod container page won't work because it requires a specific running pod. We have to create the new page.

cloudbehl commented:
> Yes, but there is no page yet to link to. The existing pod container page won't work because it requires a specific running pod. We have to create the new page.

Understood, Thanks!

cloudbehl commented:
@spadgett @zherman0 can you review this patch. or should I break this also in multiple PR's?

spadgett commented on frontend/public/components/mounted-vol.tsx:
```diff
@@ -0,0 +1,94 @@
+/* eslint-disable no-undef */
+
+export type MountedVolsProps = {
+  resourceVol: any;
```
I'd call this `podTemplate`, which makes it more clear what it really is.

```suggestion
  podTemplate: any;
```

spadgett commented on frontend/public/components/mounted-vol.tsx:
```diff
@@ -0,0 +1,94 @@
+              <div className="col-sm-3 col-xs-3">Type</div>
+              <div className="col-sm-3 hidden-3">Permissions</div>
+              <div className="col-sm-3 col-xs-3">Utilized By</div>
+            </div>
```
I'm not sure why these tables are different. We should be able to show the same things regardless of whether this is a pod. (For non-pod workloads, we simply don't make the container a link until we have a page to link to.)

spadgett replied:
OK, I see. I'd still make them the same, which keeps the UI consistent and the code simpler.

cloudbehl commented:
@spadgett @zherman0  Update pull request, please have a look.

spadgett commented on frontend/public/components/mounted-vol.tsx:
```diff
@@ -0,0 +1,65 @@
+  const name = volumeMounts.name;
+  const mountPermissions = getVolumeMountPermissions(volumeMounts);
+
+  return <React.Fragment>
```
Unnecessary `React.Fragment`

spadgett commented on frontend/public/components/mounted-vol.tsx:
```diff
@@ -0,0 +1,65 @@
+          <div className="col-sm-3 col-xs-3">Mount Path</div>
+          <div className="col-sm-2 col-xs-3">SubPath</div>
+          <div className="col-sm-3 col-xs-3">Type</div>
+          <div className="col-sm-2 hidden-xs">Permissions</div>
```
We need to add by the utilized by column. For pods, it should be a link to the container page. For other workloads, it should just be the container name (no a link).

cloudbehl commented:
@spadgett Made changes to the PR as requested, please have a look. Thanks

spadgett commented:
Sorry for not getting back sooner. LGTM. Please rebase and squash, and we can merge

spadgett commented:
jenkins rebuild

spadgett commented:
/lgtm cancel

This is causing layout issues:

<img width="767" alt="screen shot 2018-12-20 at 10 34 11 am" src="https://user-images.githubusercontent.com/1167259/50293995-ddf9d080-0442-11e9-88c2-18a2f0be9b3e.png">

spadgett commented on frontend/public/components/mounted-vol.tsx:
```diff
@@ -0,0 +1,82 @@
+      {volumeMounts.mounts.map((m, i) => <React.Fragment key={i}>
+        <div className="col-sm-2 col-xs-3">{_.get(m, 'mountPath', '-')}</div>
+        <div className="col-sm-2 col-xs-3">{_.get(m, 'subPath', '-')}</div>
+      </React.Fragment>)}
```
This seems wrong. If there are multiple mounts, the columns will no longer match the headers, correct?

cloudbehl replied:
@spadgett correct I agree the col will not match the headers if we have multiple mount path or the longer string name.

So can I hide mount path and subpath when it increases the column size and show only name, type, permission and utilized by?

spadgett replied:
Yeah I think it makes sense to only have those columns at wider screen widths.

cloudbehl replied:
Ack will do the required changes.

spadgett replied:
It looks like this is still an issue?

cloudbehl replied:
@ 150% screen size
![screenshot from 2019-01-11 20-46-31](https://user-images.githubusercontent.com/6695156/51042376-217ad100-15e2-11e9-8027-91b6126c1df2.png)
@ 100% screen size
![screenshot from 2019-01-11 20-46-08](https://user-images.githubusercontent.com/6695156/51042377-22136780-15e2-11e9-8e7a-93de8d9aa46b.png)

cloudbehl replied:
@spadgett sorry I didn't understand. Do you want to add Overflow component for the Type column also?

spadgett replied:
It's showing the wrong data in the wrong column. You're creating new columns in a loop in the middle of rending a row, so the column headers don't match the content.

cloudbehl replied:
@spadgett Ahh, got the problem. Overflow component has it's `-` if the value not there and I was also adding `-` if the value is not there. So the second hyphen was getting shifted to another row.

cloudbehl replied:
I fixed it from my side. can you test it once or can you provide me setup where I can test? I was not able to replicate the issue because I didn't have the monitoring deployed.

I tried deploying using the [prometheus-operator](https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus) but the prometheus-k8s stateful set was not created in monitoring namespace.

cloudbehl replied:
![screenshot from 2019-01-15 16-02-37](https://user-images.githubusercontent.com/6695156/51175027-35268000-18df-11e9-95ce-8cda23b03b42.png)

@spadgett Fixed the issue. Please have a look.

spadgett commented:
We need to make sure long text doesn't overlap other columns.

I think we might be trying to show more content than there is room for. Paths can be long, and we're giving it very little space.

alecmerdler commented:
jenkins rebuild

spadgett commented:
@cloudbehl Any update?

FYI, feature freeze is Friday this week (and this will likely conflict with #1043).

cloudbehl commented:
> @cloudbehl Any update?
> 
> FYI, feature freeze is Friday this week (and this will likely conflict with #1043).

@spadgett updated the PR. Sorry for the delay, was busy with the gluster related work.

spadgett commented:
> @spadgett updated the PR. Sorry for the delay, was busy with the gluster related work.

Understood. Thanks for the update! I'll take a look.

cloudbehl commented:
> @cloudbehl Are you able to rebase? master is open for 4.1 features

@spadgett rebased the PR.

spadgett commented:
@cloudbehl Thanks. Looks like you do have some linter errors after the rebase

spadgett commented:
/hold cancel

spadgett commented on frontend/public/components/mounted-vol.tsx:
```diff
@@ -0,0 +1,81 @@
+    {volumeMounts.mounts.map((m, i) => <React.Fragment key={i}>
+      <div className="row">
+        <div className="col-lg-2 col-md-3 col-sm-4 col-xs-5">{name}</div>
+        <div className="col-lg-2 col-md-3 col-sm-5 col-xs-7 co-break-word">{_.get(m, 'mountPath', '-')} </div>
```
```suggestion
        <div className="col-lg-2 col-md-3 col-sm-5 col-xs-7 co-break-any co-select-to-copy">{_.get(m, 'mountPath', '-')} </div>
```

spadgett commented on frontend/public/components/mounted-vol.tsx:
```diff
@@ -0,0 +1,81 @@
+      <div className="row">
+        <div className="col-lg-2 col-md-3 col-sm-4 col-xs-5">{name}</div>
+        <div className="col-lg-2 col-md-3 col-sm-5 col-xs-7 co-break-word">{_.get(m, 'mountPath', '-')} </div>
+        <div className="col-lg-2 col-md-2 col-sm-3 hidden-xs co-break-word">{_.get(m, 'subPath', '-')} </div>
```
```suggestion
        <div className="col-lg-2 col-md-2 col-sm-3 hidden-xs co-break-any co-select-to-copy">{_.get(m, 'subPath', '-')} </div>
```

spadgett commented on frontend/public/components/mounted-vol.tsx:
```diff
@@ -0,0 +1,81 @@
+  const mountPermissions = getVolumeMountPermissions(volumeMounts);
+
+  return <div>
+    {volumeMounts.mounts.map((m, i) => <React.Fragment key={i}>
```
If `name` is unique, it would be better to use that here. (I'm not sure offhand.)

```suggestion
    {volumeMounts.mounts.map((m) => <React.Fragment key={m.name}>
```

spadgett commented on frontend/public/components/mounted-vol.tsx:
```diff
@@ -0,0 +1,81 @@
+        <div className="col-lg-2 col-md-3 col-sm-4 col-xs-5">{name}</div>
+        <div className="col-lg-2 col-md-3 col-sm-5 col-xs-7 co-break-word">{_.get(m, 'mountPath', '-')} </div>
+        <div className="col-lg-2 col-md-2 col-sm-3 hidden-xs co-break-word">{_.get(m, 'subPath', '-')} </div>
+        <div className="col-lg-2 col-md-2 hidden-sm hidden-xs">
```
We might revisit the order we drop the columns. I feel like the kind of mount is more important than some of the other values we show at `sm` screen widths.

spadgett commented on frontend/public/components/mounted-vol.tsx:
```diff
@@ -0,0 +1,81 @@
+            <div className="col-lg-2 hidden-md hidden-sm hidden-xs">Utilized By</div>
+          </div>
+          <div className="co-m-table-grid__body">
+            {getVolumeMountsByPermissions(podTemplate).map((v, i) => <Volume key={i} volumeMounts={v} pod={podTemplate} />)}
```
`name` would be a better key here if unique.

spadgett commented:
/retest

spadgett commented:
/retest

cloudbehl commented:
@spadgett I have a question, if I commit suggestion from the button given in UI will it create a new commit or something for each suggestion you provided?

I never did it, just wanted to know if it's a right approach?

spadgett commented:
I believe it adds a new commit, but we’d want to test the changes. I’d leave well enough alone since this is tagged for merge. We can tweak in a follow on if desired

spadgett commented:
/retest

cloudbehl commented:
> I believe it adds a new commit, but we’d want to test the changes. I’d leave well enough alone since this is tagged for merge. We can tweak in a follow on if desired

Sure, I will create new PR with then with the tested changes.
